### ISSUE TITLE
[Feat]: "Metric Similarity"

### ISSUE THREAD

**Issue by andrewm4894** [netdata/netdata-cloud#793]
### Problem

Sometimes i have a line in NC and i would like to see what lines somewhere else on the page are most similar looking to it as often this can help me find some of the main "drivers" of the line i am looking at.

### Description

This feature would be quite similar in UI/UX to Metric correlations but different enough to be its own separate thing. 

There should be a "Metric similarity" button next to existing "Metric correlations" button:

![image](https://user-images.githubusercontent.com/2178292/233053927-7ad28480-a46c-491a-a0ee-381388a8e39a.png)

There should also be an entry point when i select an individual dimension - so some sort of similar flow or icon like we have for MC when you highlight an area of a chart.

![image](https://user-images.githubusercontent.com/2178292/233054466-9c6f0c72-311c-472d-99de-be845c00f384.png)

When i press either entry point i get a similar options type input section like we have for MC.

It will have some defaults prepopulated for the main input params of metric similarity

Main inputs:

- "Metric": The "Target Metric" you want to find metrics that are most similar too. 
- "Comparison Metrics": What set of metrics do you want to compare against - maybe this should just be all metrics and we could drop this input but just mentioning it here as an option.

![image](https://user-images.githubusercontent.com/2178292/233054763-5ce11bd3-7041-4936-9eb7-eef44684d223.png)

The results will be a similar filtered view of the overview tab - very similar to existing MC results - apart from in this use case by default the slide would be set to filter to just metrics with a similarity score of 90% or more. We can still have slide work like it does with show less and more by having slide just use the distance which is (1-similarity). 

So the end result here will be a filtered overview screen where all the dims should look somewhat similar to the "target metric" you started with.

"Show less" should filter to more similar looking metrics.
"Show more" should add more metrics based on similarity.

### Importance

really want

### Value proposition

1. Make it very easy for users to find what other metrics somewhere else on the page "look most similar" to the one they are looking at.

here is a demo to show how it might work:

https://user-images.githubusercontent.com/2178292/233061254-d590dc22-458c-421b-9d99-8e359ddc0d40.mp4

### Proposed implementation

- This should be a cloud service with cloud routing to a parent if possible.
- The FE will then receive a response very similar to existing MC response and that will drive same filtering type approach that MC uses.

#### Algorithm

A POC app is available here based on agent data: https://andrewm4894-netdata-ml-apps-home-uvftgr.streamlit.app/Metric_Similarity

The "recipe" for this can be seen here: https://github.com/andrewm4894/netdata-ml-apps/blob/main/pages/Metric_Similarity.py

NC would need to route the data to a netdata parent if possible or just pull all the data needed and then run the normalization step on each metric and the distance measure on the target metric and each candidate metric.

This can be done all at once or on batches of metrics if thats easier for the BE - we dont need to have all the data in one place to do this - we can do each pairwise distance measure one by one or in batches.

**Comment by andrewm4894** [comment]
@ktsaou @ralphm @cakrit @shyamvalsan @sashwathn @hugovalente-pm 

In our ideation call today @cakrit had idea to do this in a way that can be done soley on the agent and avoid needing to do any cloud aggregation or compute. 

1. NC would send the "target line" data (slice of the visible window or a highlighted subsection) to each agent.
2. Agents or parents would compute similarity to all `node,instance,dimension` level metrics on the nodes one by one. 
3. Results in NC then would be grouped by `node,instance,dimension` and overview tab would be filtered just for those metrics at the `node,instance,dimension` level that are most similar. 
4. This means that you can start the flow from anywhere and at any level - its just a line to be sent to the agent.
5. Aim of agent then is to ask if it has any lines similar on itself.
6. Results then in NC are all at `node,instance,dimension` which is typically the sort of drill in level you want to get to anyway.

Doing it this way would mean the computation could all be done on agents since results back into NC then are just similarity % 

So there could be a `/similarity` endpoint on agent that takes in:

- `target_data`
- `target_after`
- `target_before`

And just returns list of metric similarity scores for all dims on the node. 

NC then would handle all the filtering etc, change the grouping to be by node,instance,dimension and then we have results to use to filter the overview to find what similar metrics have been found anywhere within all the nodes and instances that make up the overview from where we started.

**Comment by andrewm4894** [comment]
linking: https://github.com/netdata/netdata/issues/14967

**Comment by hugovalente-pm** [comment]
> There should be a "Metric similarity" button next to existing "Metric correlations" butto
> So there could be a /similarity endpoint on agent that takes in:

These points that hint into making something new and not linked to our /weights endpoint are a bit concerning since we even have this suggested work to merge Anomaly Detection into Metric Correlations - https://github.com/netdata/netdata-cloud/issues/630

We should aim to group all these scoring methods, as you are hinting now on https://github.com/netdata/netdata/issues/14967, under the `/weights` endpoint and find a name on how we want to call it. on the above ticket discussion it seemed ["Correlations" was the winner](https://github.com/netdata/netdata-cloud/issues/630#issuecomment-1311642749)

**Comment by andrewm4894** [comment]
@hugovalente-pm i think it makes sense to add a new "Metric Similarity" button and entry point for various reasons (its a bit of a different entry point since you have to start from a "reference time series"). But i think we can have these discussions later for sure on how best to package it up in NC UI - i'd rather have different cleaner and more focused (do one thing well) product features rather than just trying for force everything into the one mega feature. 

Mainly as it kills discoverability if new features get added and users unaware - they just won't use them or discover them. Also becomes very hard to understand usage and what features are working better for users then others.

Will set up some follow on ideation calls for us all to discuss.

**Comment by shyamvalsan** [comment]
Yes, definitely worth further discussion. While I understand the distinction between similarity and correlation, I am concerned that we are giving our users a lot of options to pick from that they will not be equipped to make an informed decision on.

If Netdata is to truly be a tool that any developer or admin can use with a minimal learning curve, then the UX is even more important than the algorithm being used or how the data is queried.

**Comment by hugovalente-pm** [comment]
yes, where I was coming from is the same concern from Shyam.
If we provide a lot of options and users are not clear on what they should use in what situation it maybe too confusing

for sure worth further discussions,
